### DISCUSSION TITLE
2021-3 UVA - 10562 Undraw the Trees-解题报告（徐嘉辉）

### DISCUSSION THREAD

**Discussion by 1205240810** [CSGrandeur/s-1problem1day1ac#137] [Solutions]
## 题目大意

给你一个用字符串表示的多叉树，让你处理字符串，然后把树用括号的形式表示出来，我们的树是用括号和节点标签绘制的。每个子树都以一个开头

括号，并以右括号结束；括号内列出了节点标签。这个子树总是从左到右列出。保证所有数据，每棵树在一行，在我们的数据库中，每棵树在

一行，它们不包含除节点标签和括号对之外的任何字符。如果原始树有这样的重复，则可以重复节点标签。

## 解题思路

这道题没什么算法可言，按照题意模拟即可，我们先用字符数组存储整个完整的树，然后遍历第一行，找到树的起点，之后dfs搜索这个数的子树并且按要求依次处理下两行的起点和终点，然后继续遍历起点到终点下一行的字母，如果有子树，就按照dfs序继续遍历这个子树，在所有操作结束后输出一个‘）’。

## 复杂度分析

我们假设每行子串长度都是n，最多遍历每个字符一次，算法复杂度为O（n²）

## 算法及心得体会

需要注意的就是处理字符串时，注意每个字符串结尾和是‘\0’,还有使用gets之前需要用getchar读取掉上一行末尾的换行符\n。

## 代码

```c++
#include<stdio.h>
#include<string.h>
#include<iostream>
char a[305][305];
void dfs(int i,int u)//i是当前横坐标，u是当前纵坐标；
{
	int j,front=i,rear=i;
	printf("%c(",a[u][i]);
	if(a[u+1][i]!='|')
	{
		printf(")");
		return ;
	} 
	 	while(1)//确定二叉树起始位置 
	 	{
	 		if(a[u+2][front-1]=='-')
				front--;
			else 
				break;	
		}
		while(1)//确定二叉树的结尾 
		{
			if(a[u+2][rear+1]=='-')
				rear++;
			else 
				break;
		}
		for(j=front;j<=rear;j++)
			if(a[u+3][j]!=' '&&a[u+3][j]!='\0')
				dfs(j,u+3);
		printf(")");
} 
int main()
{
	int t,i;
	scanf("%d",&t);
	getchar();
	while(t--)
	{
		i=0;
		memset(a,' ',sizeof(a));
		while(gets(a[i])&&a[i][0]!='#')
		i++;
		printf("(");
		if(i!=0)
		{
			int sizea=strlen(a[0]);
			for(i=0;i<sizea;i++)
			if(a[0][i]!=' ')dfs(i,0);
		}
		printf(")\n");
	}
	return 0;
}
 
```

**Comment by CSGrandeur** [comment]
`getchar()` “吃掉”上一行换行不是最安全的方式，与文件编码有关（`\r\n`等情况）

用一个`gets()`代替这里的`getchar()`更保险。
